merkle/src/tree_impl.rs

```diff
@@ -0,0 +1,183 @@
+// Copyright 2020 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0
+
+use super::*;
+use std::marker::PhantomData;
+
+/// Special complete binary merkle tree that is compatible with
+/// <https://github.com/o1-labs/snarky/blob/master/src/base/merkle_tree.ml>
+/// whose leaf nodes are at the same height
+pub struct MinaMerkleTree<TItem, THash, THasher, TMerger>
+where
+    THasher: MerkleHasher<Item = TItem, Hash = THash>,
+    TMerger: MerkleMerger<Hash = THash>,
+    THash: Clone,
+{
+    depth: u32,
+    leafs: Vec<(THash, TItem)>,
+    nodes: Vec<Option<THash>>,
+
+    _pd_hasher: PhantomData<THasher>,
+    _pd_merger: PhantomData<TMerger>,
+}
+
+impl<TItem, THash, THasher, TMerger> MinaMerkleTree<TItem, THash, THasher, TMerger>
+where
+    THasher: MerkleHasher<Item = TItem, Hash = THash>,
+    TMerger: MerkleMerger<Hash = THash>,
+    THash: Clone,
+{
+    /// Creates a new instance of [MinaMerkleTree]
+    pub fn new() -> Self {
+        Default::default()
+    }
+
+    /// Creates a new instance of [MinaMerkleTree] with estimated capacity of leaves
+    pub fn with_capacity(capacity: usize) -> Self {
+        let protential_depth = calculate_depth(capacity);
+        let protential_node_count = calculate_node_count(protential_depth);
+        Self {
+            leafs: Vec::with_capacity(capacity),
+            nodes: Vec::with_capacity(protential_node_count),
+            ..Default::default()
+        }
+    }
+
+    fn clear_dirty_hashes(&mut self, leaf_index: usize) {
```

Comment:
what does this do? a comment would be good!

Reply:
Fixed